recipes-extended/xen/files/libxl-openxt-helpers.patch

```diff
@@ -113,51 +113,79 @@ PATCHES
      libxl__store_libxl_entry(gc, dm_domid, "stubdom-version",
 --- a/tools/libxl/libxl_domain.c
 +++ b/tools/libxl/libxl_domain.c
-@@ -1023,6 +1023,7 @@ void libxl__destroy_domid(libxl__egc *eg
+@@ -1020,11 +1020,15 @@ static void domain_destroy_domid_cb(libx
+                                     libxl__ev_child *destroyer,
+                                     pid_t pid, int status);
+ 
++/* OpenXT functions */
++static void kill_openxt_helpers(libxl__gc * gc, uint32_t dm_domid);
++
+ void libxl__destroy_domid(libxl__egc *egc, libxl__destroy_domid_state *dis)
+ {
+     STATE_AO_GC(dis->ao);
+     libxl_ctx *ctx = CTX;
      uint32_t domid = dis->domid;
++    uint32_t stubdom_id;
      char *dom_path;
      int rc, dm_present;
-+    char *pid;
  
-     libxl__ev_child_init(&dis->destroyer);
+@@ -1042,7 +1046,8 @@ void libxl__destroy_domid(libxl__egc *eg
  
-@@ -1072,6 +1073,33 @@ void libxl__destroy_domid(libxl__egc *eg
- 
-         libxl__qmp_cleanup(gc, domid);
+     switch (libxl__domain_type(gc, domid)) {
+     case LIBXL_DOMAIN_TYPE_HVM:
+-        if (libxl_get_stubdom_id(CTX, domid)) {
++        stubdom_id = libxl_get_stubdom_id(CTX, domid);
++        if (stubdom_id) {
+             dm_present = 0;
+             break;
+         }
+@@ -1079,6 +1084,8 @@ void libxl__destroy_domid(libxl__egc *eg
+         libxl__destroy_device_model(egc, &dis->ddms);
+         return;
+     } else {
++        /* OpenXT: if the domain has a stubdom, we kill the stubdom helpers here */
++        kill_openxt_helpers(gc, stubdom_id);
+         dm_destroy_cb(egc, &dis->ddms, 0);
+         return;
      }
-+    /* OpenXT: if the domain has a stubdom, we kill the stubdom helpers here */
-+    int dm_domid = libxl_get_stubdom_id(ctx, domid);
-+    if (dm_domid) {
-+        /* OpenXT: Kill qmp-helper */
-+        pid = libxl__xs_read(gc, XBT_NULL,
-+                             GCSPRINTF("/local/domain/%d/"XS_QMP_PID, dm_domid));
-+        if (!pid)
-+            LOG(ERROR, "Cannot get qmp_helper pid for domain %d", dm_domid);
-+        else
-+            kill(strtol(pid, NULL, 10), SIGKILL);
+@@ -1107,6 +1117,34 @@ static void dm_destroy_cb(libxl__egc *eg
+     libxl__devices_destroy(egc, &dis->drs);
+ }
+ 
++static void kill_openxt_helpers(libxl__gc *gc, uint32_t dm_domid) {
```

Comment:
Can you move this function earlier to avoid the forward declaration?

Reply:
Sure